foundations/src/telemetry/tokio/mod.rs

```diff
@@ -0,0 +1,364 @@
+//!     
+//!     monitor.add_runtime(None, Some(i), r.handle());
+//!
+//!     thread::spawn(move || r.block_on(async {
+//!         loop {
+//!             println!("Poll");
+//!             tokio::time::sleep(Duration::from_secs(1)).await;
+//!         }
+//!     }));
+//! }
+//!
+//! loop {
+//!     monitor.record_sample();
+//!
+//!     // record metrics roughly twice a second
+//!     std::thread::sleep(Duration::from_secs_f32(0.5));
+//! }
+//! ```
+
+use crate::telemetry::metrics::{metrics, Counter, Gauge};
+use std::sync::atomic::Ordering;
+use std::sync::Arc;
+use tokio::runtime::{Handle, RuntimeMetrics};
+
+/// Monitors a set of runtimes, allowing metrics to be periodically collected for each tracked runtime.
+#[derive(Default)]
+pub struct RuntimeMonitor {
+    runtimes: Vec<RuntimeHandle>,
+}
+
+impl RuntimeMonitor {
+    /// Construct a new runtime monitor which initially monitors no runtimes.
+    pub fn new() -> Self {
+        Self {
+            runtimes: Vec::new(),
+        }
+    }
+
+    /// Add a runtime to the monitor, optionally with a name and/or id in case you are monitoring multiple runtimes.
+    ///
+    /// Runtimes should be uniquely identifiable by both label and id.
+    pub fn add_runtime(
+        &mut self,
+        runtime_name: Option<Arc<str>>,
+        runtime_id: Option<usize>,
+        handle: &Handle,
+    ) {
+        if self.runtimes.iter().any(|handle| {
+            (handle.runtime_name.is_some() != runtime_name.is_some())
+                || (handle.runtime_id.is_some() != runtime_id.is_some())
+        }) {
+            panic!("If you specify a runtime name or ID for one runtime, that setting must be specified for all");
+        }
+
+        if self.runtimes.iter().any(|handle| {
+            (handle.runtime_name.is_some() && handle.runtime_name == runtime_name)
+                || (handle.runtime_id.is_some() && handle.runtime_id == runtime_id)
+        }) {
+            panic!("Any runtime names or IDs provided to the RuntimeMonitor must be unique");
+        }
+
+        self.runtimes.push(RuntimeHandle {
+            runtime_name,
+            runtime_id,
+            handle: handle.clone(),
+        });
+    }
+
+    /// Record a sample of runtime metrics for each tracked runtime.
+    pub fn record_sample(&mut self) {
+        self.runtimes
+            .iter_mut()
+            .for_each(RuntimeHandle::record_sample)
+    }
+}
+
+struct RuntimeHandle {
+    runtime_name: Option<Arc<str>>,
+    runtime_id: Option<usize>,
+    handle: Handle,
+}
+
+impl RuntimeHandle {
+    fn record_sample(&mut self) {
+        let metrics = self.handle.metrics();
+
+        self.record_global_metrics(&metrics);
+
+        for worker_idx in 0..metrics.num_workers() {
+            self.record_worker_metrics(&metrics, worker_idx);
+        }
+    }
+
+    fn record_global_metrics(&mut self, metrics: &RuntimeMetrics) {
```

Comment:
make fn name consistent with how metrics are called elsewhere. I think we can name them `general_runtime_metrics` or `core_runtime_metrics or smthg like

Reply:
Thanks, I really hate the current name.

Reply:
renamed to tokio_runtime_core